core-test/src/test/scala/org/bitcoins/core/crypto/MnemonicCodeTest.scala

```diff
@@ -0,0 +1,951 @@
+package org.bitcoins.core.crypto
+
+import org.bitcoins.core.gen.CryptoGenerators
+import org.bitcoins.core.util.BitcoinSUnitTest
+import org.scalacheck.Prop
+import org.scalatest.Assertion
+import scodec.bits.{BitVector, ByteVector, HexStringSyntax, BinStringSyntax}
+
+import scala.util.Try
+
+/**
+  * [[https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#test-vectors Trezor]]
+  * test vectors
+  */
+class MnemonicCodeTest extends BitcoinSUnitTest {
+  behavior of "MnemonicCode"
+
+  it must "generate mnemonics of the right length" in {
+    val bits128 = MnemonicCode.getEntropy128Bits
+    MnemonicCode.fromEntropy(bits128).words.length must be(12)
+
+    val bits160 = MnemonicCode.getEntropy160Bits
+    MnemonicCode.fromEntropy(bits160).words.length must be(15)
+
+    val bits192 = MnemonicCode.getEntropy192Bits
+    MnemonicCode.fromEntropy(bits192).words.length must be(18)
+
+    val bits224 = MnemonicCode.getEntropy224Bits
+    MnemonicCode.fromEntropy(bits224).words.length must be(21)
+
+    val bits256 = MnemonicCode.getEntropy256Bits
+    MnemonicCode.fromEntropy(bits256).words.length must be(24)
+  }
+
+  it must "fail to generate mnemonics from entropy of bad length" in {
+    def testEntropyFailure(entropy: BitVector): Assertion = {
+      val short = entropy.drop(1)
+      val long = entropy ++ bin"1"
+      Try(MnemonicCode.fromEntropy(short)).isFailure must be(true)
+      Try(MnemonicCode.fromEntropy(long)).isFailure must be(true)
+    }
+
+    testEntropyFailure(MnemonicCode.getEntropy128Bits)
+    testEntropyFailure(MnemonicCode.getEntropy160Bits)
+    testEntropyFailure(MnemonicCode.getEntropy192Bits)
+    testEntropyFailure(MnemonicCode.getEntropy224Bits)
+    testEntropyFailure(MnemonicCode.getEntropy256Bits)
+
+  }
+
+  it must "have serialization symmetry - bytes" in {
+    Prop.forAll(CryptoGenerators.mnemonicCode) { code =>
+      MnemonicCode.fromBytes(code.bytes) == code
+    }
+  }
+
+  it must "have serialization symmetry - entropy" in {
+    Prop.forAll(CryptoGenerators.mnemonicCode) { code =>
+      val entropy = code.toEntropy
+      val newCode = MnemonicCode.fromEntropy(entropy)
+      code == newCode
+    }
+  }
+
+  it must "fail on a seed phrase with bad checksum" in {
+    Prop.forAll(CryptoGenerators.mnemonicPhrase) { words =>
+      val lastWord = words.reverse.head
+      // there is a chance the last word in the seed phrase
```

Comment:
I don't understand what is happening here. Why would this matter?

Reply:
I think this is resolved if you do something like your bad length test and just alter the correct checksum?

Reply:
The last N bits (depending on how long your seed phrase is) is a checksum. In the case of a 12 word phrase, the last word is a checksum. So tweaking the last word should result in a failed phrase